frontend/app/services/PayPalService.scala

```diff
@@ -0,0 +1,82 @@
+package services
+
+import com.netaporter.uri.Uri.parseQuery
+import com.typesafe.scalalogging.LazyLogging
+import configuration.Config
+import controllers.PayPal.Token
+import controllers.routes
+import okhttp3.{FormBody, OkHttpClient, Request, Response}
+import play.api.mvc.RequestHeader
+
+object PayPalService extends LazyLogging {
+
+  // The parameters sent with every NVP request.
+  private val defaultNVPParams = Map(
+    "USER" -> Config.paypalUser,
+    "PWD" -> Config.paypalPassword,
+    "SIGNATURE" -> Config.paypalSignature,
+    "VERSION" -> Config.paypalNVPVersion)
+
+  // Takes a series of parameters, send a request to PayPal, returns response.
+  private def nvpRequest(params: Map[String, String]) = {
+
+    val client = new OkHttpClient()
+    val reqBody = new FormBody.Builder()
+    for ((param, value) <- defaultNVPParams) reqBody.add(param, value)
+    for ((param, value) <- params) reqBody.add(param, value)
+
+    val request = new Request.Builder()
+      .url(Config.paypalUrl)
+      .post(reqBody.build())
+      .build()
+
+    client.newCall(request).execute()
+  }
+
+  // Takes an NVP response and retrieves a given parameter as a string.
+  private def retrieveNVPParam(response: Response, paramName: String) = {
+    val responseBody = response.body().string()
+    val queryParams = parseQuery(responseBody)
+    val result = queryParams.paramMap(paramName).head
+    response.close()
```

Comment:
I thought `response.body().string()` had already closed the response by this point, but I may well be wrong 🐱.

As far as I can make out from the docs [here](https://square.github.io/okhttp/3.x/okhttp/okhttp3/ResponseBody.html), both the `string` and `bytes` methods implicitly close the response body.